File: pkg/wire/relayenvelope/relayenvelope_test.go
package relayenvelope_test

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/agentre-hub/agentre/pkg/wire/relayenvelope"
)

func TestWrapUnwrap_GivenAFrame_ThenTheChannelIDAndPayloadComeBackIntact(t *testing.T) {
	t.Parallel()

	envelope, err := relayenvelope.Wrap("c0ffee", []byte{0x01, 0x02, 0x03})
	require.NoError(t, err)

	channelID, payload, err := relayenvelope.Unwrap(envelope)
	require.NoError(t, err)
	require.Equal(t, "c0ffee", channelID)
	require.Equal(t, []byte{0x01, 0x02, 0x03}, payload)
}

// 空载荷是合法的:它是「这条通道关了」的信号。
func TestWrapUnwrap_GivenAnEmptyPayload_ThenItSurvivesAsTheChannelClosedSignal(t *testing.T) {
	t.Parallel()

	envelope, err := relayenvelope.Wrap("c0ffee", nil)
	require.NoError(t, err)

	channelID, payload, err := relayenvelope.Unwrap(envelope)
	require.NoError(t, err)
	require.Equal(t, "c0ffee", channelID)
	require.Empty(t, payload)
}

func TestWrap_GivenAnUnusableChannelID_ThenItIsRejected(t *testing.T) {
	t.Parallel()

	for name, channelID := range map[string]string{
		"empty":    "",
		"too long": strings.Repeat("a", relayenvelope.MaxChannelIDBytes+1),
	} {
		t.Run(name, func(t *testing.T) {
			t.Parallel()
			_, err := relayenvelope.Wrap(channelID, []byte{0x01})
			require.Error(t, err)
		})
	}
}

func TestWrap_GivenAChannelIDExactlyAtTheLimit_ThenItIsAccepted(t *testing.T) {
	t.Parallel()

	channelID := strings.Repeat("a", relayenvelope.MaxChannelIDBytes)
	envelope, err := relayenvelope.Wrap(channelID, []byte{0x01})
	require.NoError(t, err)
	require.LessOrEqual(t, int64(len(envelope)-1), relayenvelope.MaxEnvelopeBytes)

	got, _, err := relayenvelope.Unwrap(envelope)
	require.NoError(t, err)
	require.Equal(t, channelID, got)
}

// 中继上收到的每一帧都是**别的设备**发来的字节。三个宿主从前各写一份解析,校验各不
// 相同 —— 浏览器那份只查截断,长度 0 照收(通道 ID 成空串、整段载荷当帧交出去),
// 非法 UTF-8 静默替换成 U+FFFD。这一组用例是那三份的公约数,现在只有一份实现。
func TestUnwrap_GivenAMalformedEnvelope_ThenItIsRejected(t *testing.T) {
	t.Parallel()

	tooLong := append(
		[]byte{0x00, byte(relayenvelope.MaxChannelIDBytes + 1)},
		strings.Repeat("a", relayenvelope.MaxChannelIDBytes+1)...,
	)

	for name, envelope := range map[string][]byte{
		"shorter than the header":   {0x00},
		"declares a zero-length ID": {0x00, 0x00, 0x01, 0x02},
		"truncated before the ID":   {0x00, 0x08, 'a', 'b'},
		"declares an oversized ID":  tooLong,
		"channel ID is not UTF-8":   {0x00, 0x02, 0xff, 0xfe, 0x01},
	} {
		t.Run(name, func(t *testing.T) {
			t.Parallel()
			_, _, err := relayenvelope.Unwrap(envelope)
			require.Error(t, err)
		})
	}
}

File: frontend/packages/agentre-wire/src/__tests__/relay-envelope.test.ts
import { describe, expect, it } from "vitest";

import {
  MAX_CHANNEL_ID_BYTES,
  unwrapEnvelope,
  wrapEnvelope,
} from "../relay-envelope";

const bytes = (...values: number[]) => new Uint8Array(values);

describe("中继通道信封", () => {
  it("套上再拆开，通道 ID 与载荷原样回来", () => {
    const { channelId, frame } = unwrapEnvelope(
      wrapEnvelope("c0ffee", bytes(1, 2, 3)),
    );
    expect(channelId).toBe("c0ffee");
    expect(Array.from(frame)).toEqual([1, 2, 3]);
  });

  // 空载荷是合法的：它是「这条通道关了」的信号。
  it("空载荷活着穿过，它是通道关闭的信号", () => {
    const { channelId, frame } = unwrapEnvelope(
      wrapEnvelope("c0ffee", new Uint8Array(0)),
    );
    expect(channelId).toBe("c0ffee");
    expect(frame.length).toBe(0);
  });

  it.each([
    ["空", ""],
    ["超长", "a".repeat(MAX_CHANNEL_ID_BYTES + 1)],
  ])("拒绝%s的通道 ID", (_name, channelId) => {
    expect(() => wrapEnvelope(channelId, bytes(1))).toThrow();
  });

  it("刚好顶格的通道 ID 收下", () => {
    const channelId = "a".repeat(MAX_CHANNEL_ID_BYTES);
    expect(unwrapEnvelope(wrapEnvelope(channelId, bytes(1))).channelId).toBe(
      channelId,
    );
  });

  // 中继上收到的每一帧都是**别的设备**发来的字节。这一组是 Go 那侧
  // relayenvelope.Unwrap 的同一张表：从前这里只查截断，另外三条一条都不查。
  it.each([
    ["比信封头还短", bytes(0)],
    ["自报长度为 0", bytes(0, 0, 1, 2)],
    ["在通道 ID 中间被截断", bytes(0, 8, 97, 98)],
    [
      "自报长度超上限",
      new Uint8Array([
        0,
        MAX_CHANNEL_ID_BYTES + 1,
        ...new Array(MAX_CHANNEL_ID_BYTES + 1).fill(97),
      ]),
    ],
    ["通道 ID 不是合法 UTF-8", bytes(0, 2, 0xff, 0xfe, 1)],
  ])("拒绝%s的信封", (_name, envelope) => {
    expect(() => unwrapEnvelope(envelope)).toThrow();
  });
});
